owncloudApp/src/main/java/com/owncloud/android/presentation/ui/security/PatternActivity.kt

```diff
@@ -0,0 +1,322 @@
+            if (patternViewModel.checkPatternIsValid(patternValue)) {
+                patternViewModel.removePattern()
+                val result = Intent()
+                setResult(RESULT_OK, result)
+                patternError.visibility = View.INVISIBLE
+                notifyDocumentProviderRoots(applicationContext)
+                finish()
+            } else {
+                showErrorAndRestart(
+                    R.string.pattern_incorrect_pattern,
+                    R.string.pattern_enter_pattern, View.INVISIBLE
+                )
+            }
+        } else if (ACTION_REQUEST_WITH_RESULT == intent.action) {
+            //This block is executed when the user is setting the pattern lock (i.e enabling the pattern lock)
+            if (!confirmingPattern) {
+                patternError.visibility = View.INVISIBLE
+                requestPatternConfirmation()
+            } else if (confirmPattern()) {
+                savePatternAndExit()
+            } else {
+                showErrorAndRestart(
+                    R.string.pattern_not_same_pattern,
+                    R.string.pattern_enter_pattern, View.VISIBLE
+                )
+            }
+        }
+    }
+
+    private fun showErrorAndRestart(
+        errorMessage: Int, headerMessage: Int,
+        explanationVisibility: Int
+    ) {
+        patternValue = null
+        patternError.setText(errorMessage)
+        patternError.visibility = View.VISIBLE
+        patternHeader.setText(headerMessage)
+        patternExplanation.visibility = explanationVisibility
+    }
+
+    /**
+     * Ask to the user to re-enter the pattern just entered before saving it as the current pattern.
+     */
+    protected fun requestPatternConfirmation() {
+        patternLockView.clearPattern()
+        patternHeader.setText(R.string.pattern_reenter_pattern)
+        patternExplanation.visibility = View.INVISIBLE
+        confirmingPattern = true
+    }
+
+    protected fun confirmPattern(): Boolean {
+        confirmingPattern = false
+        return newPatternValue != null && newPatternValue == patternValue
+    }
+
+    private fun savePatternAndExit() {
+        val result = Intent()
+        patternViewModel.setPattern(patternValue!!)
+        setResult(RESULT_OK, result)
+        notifyDocumentProviderRoots(applicationContext)
+        finish()
+    }
+
+    override fun onSaveInstanceState(outState: Bundle) {
+        super.onSaveInstanceState(outState)
+        outState.apply {
+            putBoolean(KEY_CONFIRMING_PATTERN, confirmingPattern)
+            putString(KEY_PATTERN_STRING, patternValue)
+            putString(PATTERN_HEADER_VIEW_TEXT, patternHeader.text.toString())
+            if (patternExplanation.visibility == View.VISIBLE) {
+                putBoolean(PATTERN_EXP_VIEW_STATE, true)
+            } else {
+                putBoolean(PATTERN_EXP_VIEW_STATE, false)
+            }
```

Comment:
You can use `isVisible` in those cases to check that the `visibility == VISIBLE`
```suggestion
            putBoolean(PATTERN_EXP_VIEW_STATE, patternExplanation.isVisible)
```

What do you think? Maybe you can use this on other places 👍